### PR TITLE
yamlgraph: Implement easy writing resource structure

### PR DESCRIPTION
**Pull request by mildred**
Discussed in #141

Add a version tag to the yaml graph. Version 2 has a new structure where
resources are no longer grouped by their kind. Resources can also reference
other resources in a before and after list, which will create edges. It is
no longer necessary to have the edges specified separately.

Version 1 (or no version) is still compatible, though it has been
refactored to avoid use of the reflect package, and use an extensible list
of registred resource kinds. This also has the benefit of removing the
empty VirtRes and AugeasRes struct types when compiling without libvirt and
libaugeas.

### CODE REVIEW

File: resources/resources.go [hunk 1]
```diff
@@ -39,6 +39,20 @@ import (
 	"golang.org/x/time/rate"
 )
 
+var registredResources map[string]func() Res = map[string]func() Res{}
```
Comment (purpleidea) [inline resources/resources.go]: typo registered*

**Comment by mildred** [comment]
I completely understand, especially since you explicitly said you didn't really want to focus on YAML right now.

**Comment by purpleidea** [comment]
On Tue, Feb 14, 2017 at 3:37 PM, mildred <notifications@github.com> wrote:
>
> I completely understand, especially since you explicitly said you didn't really want to focus on YAML right now.

This isn't a NACK, rather a "blocked until feat/mgmt gets merged"
(hopefully this week) and a "needs ACK/review" from someone else. Ping
in channel for a review... :)

**Comment by roidelapluie** [comment]
I will review that later this week. This is big.

**Comment by mildred** [comment]
After using this a few days, It may still not be ready to merge. Changing the format should not be done lightly. I still think an improved YAML format could be grat to have, a sort of mid-ground between the graph dump we currently have and the full-fledged DSL. But I am no longer sure about this version of the format I wrote.

Still, there are things included in this PR that I think would be great to include. The part that parses the current format differently, getting rid of the virt_disabled.go and augead_disabled.go files, and stop using the reflect package for parsing.

Basically, it boils down to the code in `gconfig.go` in this PR, especially:

- `func (c *GraphConfig) UnmarshalYAML(...)`
- `func (r *ResourceV1) UnmarshalYAML(...)`
- `func (r *ResourceV1) Decode(kind string) (err error)`

This could be presented as a separate PR, and I think it could help with conditional compilation of augeas and libvirt.

**Comment by purpleidea** [comment]
@mildred Okay, I had another quick look, and while I don't want to merge all of this, there are a few parts I would like to look into:

1) Using the resources without the explicit list and the reflection. That might be nice on its own.

2) Implementing this as a different GAPI. This means it can stand alone without needing to have a V1 vs. V2 flag. The API: https://github.com/purpleidea/mgmt/blob/master/gapi/gapi.go#L44 Yamlgraph example: https://github.com/purpleidea/mgmt/blob/master/yamlgraph/gapi.go#L31

(You could have like yamlgraphv2 or something...)

3) Lastly, the refactor I did of Init which isn't merge in git master needs to go in first :( Sorry about that. I hate when I'm the blocker.

The first two points can be worked on if you're willing. Having the hardcoded list in particular is a dumb thing that I did.

**Comment by mildred** [comment]
This must be closed in favor of #163. Just have to find time for it (and for real this time)
